// @loopback/testlab/dist/src/shot.js
"use strict";
// Copyright IBM Corp. 2017,2018. All Rights Reserved.
// Node module: @loopback/testlab
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT
Object.defineProperty(exports, "__esModule", { value: true });
const util = require("util");
const shot_1 = require("shot");
exports.inject = shot_1.inject;
const express = require("express");
// tslint:disable-next-line:variable-name
const ShotRequest = require('shot/lib/request');
function stubServerRequest(options) {
    const stub = new ShotRequest(options);
    // Hacky workaround for Express, see
    // https://github.com/expressjs/express/blob/4.16.3/lib/middleware/init.js
    // https://github.com/hapijs/shot/issues/82#issuecomment-247943773
    // https://github.com/jfhbrook/pickleback
    Object.assign(stub, ShotRequest.prototype);
    return stub;
}
exports.stubServerRequest = stubServerRequest;
// tslint:disable-next-line:variable-name
const ShotResponse = require('shot/lib/response');
function stubServerResponse(request, onEnd) {
    const stub = new ShotResponse(request, onEnd);
    // Hacky workaround for Express, see
    // https://github.com/expressjs/express/blob/4.16.3/lib/middleware/init.js
    // https://github.com/hapijs/shot/issues/82#issuecomment-247943773
    // https://github.com/jfhbrook/pickleback
    Object.assign(stub, ShotResponse.prototype);
    return stub;
}
exports.stubServerResponse = stubServerResponse;
function stubHandlerContext(requestOptions = { url: '/' }) {
    const request = stubServerRequest(requestOptions);
    let response;
    let result = new Promise(resolve => {
        response = new ShotResponse(request, resolve);
    });
    const context = { request, response: response, result };
    defineCustomContextInspect(context, requestOptions);
    return context;
}
exports.stubHandlerContext = stubHandlerContext;
function stubExpressContext(requestOptions = { url: '/' }) {
    const app = express();
    const request = new ShotRequest(requestOptions);
    // mix in Express Request API
    const RequestApi = express.request;
    for (const key of Object.getOwnPropertyNames(RequestApi)) {
        Object.defineProperty(request, key, Object.getOwnPropertyDescriptor(RequestApi, key));
    }
    request.app = app;
    request.originalUrl = request.url;
    parseQuery(request);
    let response;
    let result = new Promise(resolve => {
        response = new ShotResponse(request, resolve);
        // mix in Express Response API
        Object.assign(response, express.response);
        const ResponseApi = express.response;
        for (const key of Object.getOwnPropertyNames(ResponseApi)) {
            Object.defineProperty(response, key, Object.getOwnPropertyDescriptor(ResponseApi, key));
        }
        response.app = app;
        response.req = request;
        request.res = response;
    });
    const context = { app, request, response: response, result };
    defineCustomContextInspect(context, requestOptions);
    return context;
}
exports.stubExpressContext = stubExpressContext;
/**
 * Use `express.query` to parse the query string into `request.query` object
 * @param request Express http request object
 */
function parseQuery(request) {
    // Use `express.query` to parse the query string
    // See https://github.com/expressjs/express/blob/master/lib/express.js#L79
    // See https://github.com/expressjs/express/blob/master/lib/middleware/query.js
    express.query()(request, {}, () => { });
}
function defineCustomContextInspect(context, requestOptions) {
    // Setup custom inspect functions to make test error messages easier to read
    const inspectOpts = (depth, opts) => util.inspect(requestOptions, opts);
    defineCustomInspect(context.request, (depth, opts) => `[RequestStub with options ${inspectOpts(depth, opts)}]`);
    defineCustomInspect(context.response, (depth, opts) => `[ResponseStub for request with options ${inspectOpts(depth, opts)}]`);
    context.result = context.result.then(r => {
        defineCustomInspect(r, (depth, opts) => `[ObservedResponse for request with options ${inspectOpts(depth, opts)}]`);
        return r;
    });
}
function defineCustomInspect(obj, inspectFn) {
    obj[util.inspect.custom] = inspectFn;
}
