// @loopback/boot/dist/src/booters/service.booter.js
"use strict";
// Copyright IBM Corp. 2018. All Rights Reserved.
// Node module: @loopback/boot
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@loopback/core");
const context_1 = require("@loopback/context");
const base_artifact_booter_1 = require("./base-artifact.booter");
const keys_1 = require("../keys");
/**
 * A class that extends BaseArtifactBooter to boot the 'DataSource' artifact type.
 * Discovered DataSources are bound using `app.controller()`.
 *
 * Supported phases: configure, discover, load
 *
 * @param app Application instance
 * @param projectRoot Root of User Project relative to which all paths are resolved
 * @param [bootConfig] DataSource Artifact Options Object
 */
let ServiceBooter = class ServiceBooter extends base_artifact_booter_1.BaseArtifactBooter {
    constructor(app, projectRoot, serviceConfig = {}) {
        super(projectRoot, 
        // Set DataSource Booter Options if passed in via bootConfig
        Object.assign({}, exports.ServiceDefaults, serviceConfig));
        this.app = app;
        this.serviceConfig = serviceConfig;
    }
    /**
     * Uses super method to get a list of Artifact classes. Boot each file by
     * creating a DataSourceConstructor and binding it to the application class.
     */
    async load() {
        await super.load();
        this.serviceProviders = this.classes.filter(isServiceProvider);
        /**
         * If Service providers were discovered, we need to make sure ServiceMixin
         * was used (so we have `app.serviceProvider()`) to perform the binding of a
         * Service provider class.
         */
        if (this.serviceProviders.length > 0) {
            if (!this.app.serviceProvider) {
                console.warn('app.serviceProvider() function is needed for ServiceBooter. You can add ' +
                    'it to your Application using ServiceMixin from @loopback/service-proxy.');
            }
            else {
                this.serviceProviders.forEach(cls => {
                    this.app.serviceProvider(cls);
                });
            }
        }
    }
};
ServiceBooter = __decorate([
    __param(0, context_1.inject(core_1.CoreBindings.APPLICATION_INSTANCE)),
    __param(1, context_1.inject(keys_1.BootBindings.PROJECT_ROOT)),
    __param(2, context_1.inject(`${keys_1.BootBindings.BOOT_OPTIONS}#services`)),
    __metadata("design:paramtypes", [Object, String, Object])
], ServiceBooter);
exports.ServiceBooter = ServiceBooter;
/**
 * Default ArtifactOptions for DataSourceBooter.
 */
exports.ServiceDefaults = {
    dirs: ['services'],
    extensions: ['.service.js'],
    nested: true,
};
function isServiceProvider(cls) {
    const hasSupportedName = cls.name.endsWith('Provider');
    const hasValueMethod = 'value' in cls.prototype;
    return hasSupportedName && hasValueMethod;
}
